Update pulse instruction interface to the new pulse model

TsafrirA opened the pull request:
### Summary
The Pulse Instructions interface is updated to accommodate the new pulse model (#10694).

### Details and comments
Following discussions in #11464 (which is replaced by this PR), it was decided that the new pulse model will be introduced into the instructions interface in the following way.

The signature of instruction initialization will be updated to (demonstrated here for the `Play` instruction, but the same logic holds for all relevant instructions):

```python
def __init__(self, pulse: Pulse, *, target = None, frame = None, mixed_frame = None, channel = None, name = None):
```
Upon initialization, the input arguments are validated to guarantee that a correct combination is provided (mixed frame for Play, mixed frame\frame for phase\frequency instructions, etc.). This signature is the most flexible one, while also providing a path for a minimally breaking transition from the old model to the new model. An intermediate signature during the deprecation period (Qiskit 1.x) would be:

```python
def __init__(self, pulse: Pulse, channel = None, name = None, *, target = None, frame = None, mixed_frame = None):
```
such that a minimal change is needed to update existing code.

As discussed in #11464, this option is not without limitations, and it is quite possible that eventually the interface would change. However, with its flexibility the option presented here provides a solid starting step, which could be easily modified later on.

For posterity, it should be mentioned that another option which was favorably considered is:
```python
def __init__(self, pulse: Pulse, mixed_frame: Union[PulseChannel, MixedFrame, Tuple[PulseTarget, Frame]], name = None):
```

The presented implementation removes the `_validate` abstract method. The method was implemented individually for all instructions, and thus didn't reduce code, and hid the errors away from the initialization function. However, a `FrameInstruction` subclass was introduced with a `_validate` function, because the validation logic is shared across all set\shift phase\frequency instructions.

The following instructions were not changed:
- Reference
- Snapshot
- Directives

The `Acquire` instruction was updated in a shallow way, and requires further discussion. Main questions include - 
- Where should the `MemorySlot` (`RegisterSlot`) class sit in the new model?
- Should we support acquiring of a general port, or only `Qubit`?
- As a side question, is there a need for `RegisterSlot`? Does it serve any function on any relevant backend?

TsafrirA committed:
Update instruction interface

TsafrirA committed:
Some more test updates

TsafrirA committed:
Qubit ref fix

TsafrirA committed:
Hopefully avoid circular references.

nkanazawa1989 commented on qiskit/pulse/instructions/acquire.py:
```diff
@@ -40,18 +41,24 @@ class Acquire(Instruction):
-        kernel: Kernel | None = None,
+        duration: Union[int, ParameterExpression],
+        *,
+        qubit: Optional[model.Qubit] = None,
```
```suggestion
        qubit: model.Qubit | None = None,
```
Since we have `from __future__ import annotations`, you don't need to import Union and Optional from typing module. This change was made by https://github.com/Qiskit/qiskit/pull/8268 with considerable effort, and we don't need to switch back to the old convention. Same for others.

nkanazawa1989 commented on qiskit/pulse/instructions/acquire.py:
```diff
@@ -60,38 +67,50 @@ def __init__(
+
     @property
-    def channel(self) -> AcquireChannel:
+    def channel(self) -> Union[AcquireChannel, None]:
```
Can we deprecate channel and acquire at minimum cost?

TsafrirA replied:
Once we're done, I think yes.

nkanazawa1989 commented on qiskit/pulse/instructions/delay.py:
```diff
@@ -36,27 +39,72 @@ class Delay(Instruction):
+                raise PulseError(f"Expected a PulseTarget, got {target} instead.")
+            if not isinstance(frame, Frame):
+                raise PulseError(f"Expected a Frame, got {frame} instead.")
+            inst_target = MixedFrame(target, frame)
```
```suggestion
        if (channel is not None) + (mixed_frame is not None) + (target is not None) != 1:
            raise PulseError("Exactly one of mixed_frame, channel or target must be provided")

        if frame is not None:
            if not isinstance(target, PulseTarget):
                raise PulseError(f"Expected a PulseTarget, got {target} instead.")
            if not isinstance(frame, Frame):
                raise PulseError(f"Expected a Frame, got {frame} instead.")
            inst_target = MixedFrame(target, frame)
```
Equivalent right? Maybe we can delegate typecheck of `frame` and `target` to the `MixedFrame` constructor since this pattern also appears in `Play` instruction.

TsafrirA replied:
Yes, this seems equivalent.

nkanazawa1989 commented on qiskit/pulse/instructions/instruction.py:
```diff
@@ -267,3 +257,79 @@ def __repr__(self) -> str:
+        return inst_target
+
+    @property
+    def channel(self) -> Union[PulseChannel, None]:
```
Because we will deprecate channel at some point and move to mixed-frame model, we don't need to add `channel` property to newly introduced classes.

TsafrirA replied:
This isn't really an addition. It's just a way to keep the `channel` property of all frame instructions with less code clutter. It will make it easier to deprecate it later.

nkanazawa1989 replied:
I would add channel property to each subclass rather in the baseclass. In this way newly added instructions will not gain legacy property. Anyways this is small difference and doesn't block merge.

nkanazawa1989 commented on qiskit/pulse/instructions/instruction.py:
```diff
@@ -267,3 +257,79 @@ def __repr__(self) -> str:
+    ``FrameInstruction`` should be considered abstract, and should not be instantiated as is.
+    """
+
+    def _validate_and_format_frame(
```
Perhaps you also need to overwrite `_validate` in the sueprclass? Since `Instruction._validate` is called inside the constructor and it checks if all `self.channels` are `Channel` instance. When either mixed_frame or frame is passed, self.channel returns `None` and the validation will fail.

nkanazawa1989 commented on qiskit/pulse/instructions/frequency.py:
```diff
@@ -13,120 +13,111 @@
         frequency: Union[float, ParameterExpression],
-        channel: PulseChannel,
+        *,
+        frame: Frame = None,
```
```suggestion
        frame: Frame | None = None,
```
and import `from __future__ import annotations`

nkanazawa1989 commented on qiskit/pulse/instructions/phase.py:
```diff
@@ -43,56 +43,45 @@ class ShiftPhase(Instruction):
-        """Instantiate a shift phase instruction, increasing the output signal phase on ``channel``
-        by ``phase`` [radians].
+        """Instantiate a shift phase instruction, increasing the output signal phase on a given
+        mixed frame (=channel) or frame by ``phase`` [radians]..
```
```suggestion
        mixed frame (=channel) or frame by ``phase`` [radians].
```

nkanazawa1989 commented on qiskit/pulse/instructions/play.py:
```diff
@@ -14,64 +14,104 @@
+                raise PulseError(f"Expected a Frame, got {frame} instead.")
+            mixed_frame = MixedFrame(target, frame)
+        else:
+            mixed_frame = mixed_frame or channel
```
```suggestion
        if (channel is not None) + (mixed_frame is not None) + (target is not None) != 1:
            raise PulseError(
                "Exactly one of mixed_frame, channel or the duo target and frame has to be provided"
            )

        if target is not None:
            if not isinstance(target, PulseTarget):
                raise PulseError(f"Expected a PulseTarget, got {target} instead.")
            if not isinstance(frame, Frame):
                raise PulseError(f"Expected a Frame, got {frame} instead.")
            mixed_frame = MixedFrame(target, frame)
        else:
            mixed_frame = mixed_frame or channel
```
Also equivalent right?

TsafrirA replied:
This will actually fail to raise an error if you provide `frame` and `mixed_frame` (the frame argument will be ignored).

nkanazawa1989 commented on qiskit/pulse/instructions/reference.py:
```diff
@@ -53,19 +53,14 @@ def __init__(self, name: str, *extra_keys: str):
+            PulseError: When ``name`` or an ``extra_key`` contains the scope delimiter.
         """
-        for key in self.ref_keys:
+        ref_keys = (name,) + tuple(extra_keys)
```
Why did you move validation to the constructor? `_validate` is a base class method which is called after class is instantiated, which defines an extra validation protocol for instruction-specific operands. This makes constructor bit cleaner. If you don't like it, feel free to remove it from the base class (because this is just a protected method). Since you added lots of typecheck directly in the constructor, this delayed validation mechanism doesn't fit in with new framework.

nkanazawa1989 commented on test/python/pulse/test_instructions.py:
```diff
@@ -121,29 +238,111 @@ def test_operator_delay(self):
+
+        # Invalid Combinations
+        with self.assertRaises(exceptions.PulseError):
+            instructions.SetPhase(0.5, target=target)
```
This test is tied to the implementation of `SetPhase`. You need to define a clean test class within the test case which is independent of existing subclasses to be tested. Something like

```python
class TestInstruction(FrameInstruction):
    def __init__(
        *,
        frame: Frame = None,
        target: PulseTarget = None,
        mixed_frame: MixedFrame = None,
        channel: PulseChannel = None,
    ):
        ...
```

nkanazawa1989 commented on test/python/pulse/test_instructions.py:
```diff
@@ -285,23 +606,90 @@ def setUp(self):
-            instructions.Play(self.pulse_op, channels.AcquireChannel(0))
+            instructions.Play(self.pulse_op, channel=channels.AcquireChannel(0))
+
+    def test_play_arguments_validation(self):
```
I see this pattern everywhere. I personally prefer having a single validation per test. What do you think of using following pattern?

```python
from ddt import ddt, named_data, unpack

@ddt
class  ...
    @named_data(
        ["only_frame", None, None, QubitFrame(0), None],
        ["only_target", None, Qubit(0), None, None],
        ...
    )
    @unpack
    def test_invalid_construction(self, channel, target, frame, mixed_frame):
        with self.assertRaises(exceptions.PulseError):
            instructions.Play(
                self.pulse_op, 
                channel=channel, 
                target=target, 
                frame=frame, 
                mixed_frame=mixed_frame,
            )
    
```

This pattern is almost equivalent to defining a dedicated test function for each condition. `named_data` decorator is really developer friendly since it tells you the test condition in human readable manner.

TsafrirA replied:
Nifty trick!

nkanazawa1989 reviewed (changes requested):
Thanks @TsafrirA for your hard work of unittest update. I like new pattern for instruction construction. Since new model allows multiple patterns to specify how the instruction is executed, user must explicitly give operands. Although new pattern looks complicated, I believe our target users are familiar with the concept and they can use new model with well-written migration guide (TODO).

On the other hand, tons of changes to the unittests is an indication of breaking API change. In some minor release of 1.0 we should add deprecation warning for the use of non-keyword args. This is just a reminder and doesn't block this PR.

TsafrirA committed:
Corrections after review

TsafrirA commented:
Thanks @nkanazawa1989 for the quick review. I've modified the tests and some of the validation logic as you requested.

As I wrote, I removed entirely the `_validate` method. I think it wasn't very useful, as most instructions implemented it individually anyway. I think the current implementation better reflects the errors to the user, where the user wants to see them.
The only exception is the validation method which was added to the `FrameInstruction` class, because here we can really reduce code clutter.

nkanazawa1989 commented on qiskit/pulse/instructions/instruction.py:
```diff
@@ -267,3 +257,75 @@ def __repr__(self) -> str:
         )
+
+
+class FrameInstruction(Instruction, ABC):
```
Better class name would be `FrameUpdate` or `FrameChange`? (without -Instruction like `Directive` class).  FrameInstruction sounds bit vague to me because it reminds me of something _only_ takes frame, but actual implementation is not. I think our intention for its subclass must be an instruction that updates the frame value, right?

TsafrirA replied:
Sounds good.

nkanazawa1989 reviewed (commented):
One more comment for class name. Overall the PR looks good to go. Can you resolve conflict too?

TsafrirA committed:
Change FrameInstruction to FrameUpdate

TsafrirA committed:
Merge branch 'feature/pulse-ir' into InstructionInterfaceV2

TsafrirA committed:
update to reflect qiskit.test removal

TsafrirA committed:
Test update

TsafrirA commented:
@nkanazawa1989 
I've resolved the conflicts.

nkanazawa1989 reviewed (approved):
Thanks @TsafrirA LGTM :)
